jaxrs-api/src/main/java/jakarta/ws/rs/ext/Part.java

```diff
@@ -0,0 +1,219 @@
+ *         Part.withName("name1").fileName("file2.doc").content(stream2).build(),
+ *         Part.withName("name1").fileName("file3.doc").content(stream3).build());
+ * Entity entity = Entity.entity(parts, MediaType.MULTIPART_FORM_DATA);
+ * Response r = target.request().post(entity);
+ * </pre>
+ * 
+ * Note that when building a Part, the name and content are required.
+ * Other properties such as headers, file name, and media type are optional.
+ * 
+ * It is the responsibility of the implementation code to close the content input streams when sending the multipart
+ * content. Closing the stream before the implementation has sent it could result in unexpected exceptions. It is the
+ * responsibility of the calling code to close the stream when receiving the multipart content.
+ * 
+ * @since 3.1
+ */
+public interface Part {
+
+    /**
+     * Creates a new {@code Part.Builder} instance.
+     * 
+     * @param partName name of the part to create within the multipart entity
+     * @return {@link Builder} for building new {@link Part} instances
+     */
+    static Builder withName(String partName) {
+        return RuntimeDelegate.getInstance().createPartBuilder(partName);
+    }
+
+    /**
+     * Creates a new {@code Part.Builder} instance that sets the part {@code name} and
+     * {@code fileName} to the passed in {@code partAndFileName} value.
+     * <p>
+     * Logically, this is the same as {@code Part.withName(x).fileName(x)}.
+     * </p>
+     * @param partAndFileName name and filename of the part to create within the multipart entity
+     * @return {@link Builder} for building new {@link Part} instances
+     */
+    static Builder withFileName(String partAndFileName) {
+        return RuntimeDelegate.getInstance().createPartBuilder(partAndFileName).fileName(partAndFileName);
+    }
+
+    /**
+     * Returns the name of this part within the multipart entity. This will be the 
+     * "name" attribute of the {@code Content-Disposition} header for this part.
+     * 
+     * @return the part name
+     */
+    String getName();
+
+    /**
+     * Returns the filename of this part. This will be the "filename" attribute of
+     * the {@code Content-Disposition} header for this part. A filename is not
+     * required in a part, so if a filename is not present it will return
+     * {@code Optional.empty()}
+     * 
+     * @return an {@code Optional<String>} indicating the filename if present
+     */
+    Optional<String> getFileName();
+
+    /**
+     * Returns the input stream for this part. This is the content body of the part
+     * and is accessed as a stream to avoid loading potentially large amounts of
+     * data into the heap.
+     * 
+     * It is the responsibility of the calling code to close this stream after
+     * receiving it.
+     * 
+     * @return an {@code InputStream} representing the content of this part
+     */
+    InputStream getContent();
```

Comment:
I wonder if it is really sufficient that the content is only available as an `InputStream`. What about adding something like:

    <T> T getContent(Class<T> type);

This method could use the available MBRs to convert the content to all kind of types like `byte[]`, `java.io.Reader`, `javax.json.JsonObject`, etc. Or am I missing something. AFAIK Jersey supports this as well.

Of course the argument that `InputStream` prevents loading much data into the heap is valid, but that's actually the case for all resources methods which consume the request body without using an `InputStream`, no?

Also, maybe `getEntity()` would be a better name for the method?

Reply:
In fact we cannot name it `getEntity()` as the entity is the container holding *all* parts.

Reply:
Well, the term *entity* isn't strictly defined in the spec, AFAIK. But a few related terms are:

There is the *request entity* and the *response entity* for which there can only be at most one per request/response. Then we have the *entity parameter* which basically allows the resource method to access the *request entity* created from the *request entity body*.

For multipart requests we also have a single *request entity*. But this entity actually represents one or more *parts* and each one is quite similar to a *request entity* in a sense that we have a body and a media type. Maybe we could call it *part entity* or something like that in the API docs.

Reply:
Actually the term "entity" is pretty clearly defined by the HTTP spec, but not even mentioned by the specs for multiform, so we should definitively *not* mix things here, but stick with the term "content" which the multiform specs actually comes up with.

Reply:
> Actually the term "entity" is pretty clearly defined by the HTTP spec

Thanks for the hint. I wasn't aware that the term "entity" is actually from the HTTP spec. However, I still like the idea that a part also has an "entity" because it is actually a very similar concept. Jersey uses the term entity as well in the [BodyPart](https://eclipse-ee4j.github.io/jersey.github.io/apidocs/latest/jersey/org/glassfish/jersey/media/multipart/BodyPart.html) class. However, I would love to hear other thoughts as well.

Reply:
Calling the content of a part an "entity" is confusing IMO for the reasons already given here.

Reply:
Ok. Beside the name change suggested, what about my other comment?

> I wonder if it is really sufficient that the content is only available as an `InputStream`. What about adding something like:
> 
> ```
> <T> T getContent(Class<T> type);
> ```
> 
> This method could use the available MBRs to convert the content to all kind of types like `byte[]`, `java.io.Reader`, `javax.json.JsonObject`, etc. Or am I missing something. AFAIK Jersey supports this as well.
> 
> Of course the argument that `InputStream` prevents loading much data into the heap is valid, but that's actually the case for all resources methods which consume the request body without using an `InputStream`, no?

Reply:
I second that. InputStream is fine, but it should not be mandatory -- even if the multiforms specs only talk about "files".

Reply:
Yeah, I think that could work - and I think Santiago mentioned something similar.  Would we also need a separate method to handle `GenericType` parameters to convert to generic types? Something like this? `<T> T getContent(GenericType<T> type);`

I'll go ahead and start coding up both methods - but let me know if the generic type approach is unnecessary (I think we'd need it though to account for type erasure).

Reply:
Additional `getContent` methods are in.  Please take a look and let me know if there's anything in those methods that I should change.  Thanks again!

Reply:
Thanks a lot @andymc12. And sorry for my delayed response.

> Would we also need a separate method to handle `GenericType` parameters to convert to generic types? Something like this? `<T> T getContent(GenericType<T> type);`

Good question. Not sure TBH. But I guess we would need it to supported more advanced use cases like nested multipart entities!? For example:

```java
List<Part> children = part.getContent( new GenericType<List<Part>>() {} );
```

Reply:
Are nested multipart's even a thing? Seems like a can of worms. Let's keep things simple for a first version of the API.

Reply:
No, nested multiparts doesn't make much sense. This was just an example for using generics with `getContent`.

But @andymc12 already added `T getContent(GenericType<T> type)` below, so I guess the issue regarding support for generic types is already resolved.

Reply:
Right, I added the `GenericType` support more for collection use cases than for nested multiparts.  So instead of @chkal's example, I would use an example like this:
```
List<String> lines = part.getContent(new GenericType<List<String>>(){});
```

or:
```
Set<Widget> orderForm = part.getContent(new GenericType<Set<Widget>>(){});
```